mediacloud/mediawords/tm/fetch_topic_tweets.py

```diff
@@ -0,0 +1,443 @@
+        'publish_date': '2017-01-01',
+        'twitter_user': ch_post['tweet']['user']['screen_name']
+    }
+
+    topic_tweet = db.create('topic_tweets', topic_tweet)
+
+    urls_inserted = {}
+    for url_data in ch_post['tweet']['entities']['urls']:
+
+        url = url_data['expanded_url']
+
+        if url in urls_inserted:
+            next
+
+        urls_inserted[url] = True
+
+        db.create(
+            'topic_tweet_urls',
+            {
+                'topic_tweets_id': topic_tweet['topic_tweets_id'],
+                'url': url[0:1024]
+            })
+
+
+def _fetch_tweets_for_day(
+        db: DatabaseHandler,
+        twitter_class: AbstractTwitter,
+        topic: dict,
+        topic_tweet_day: dict,
+        max_tweets: int=None) -> None:
+    """
+    Fetch tweets for a single day.
+
+    If tweets_fetched is false for the given topic_tweet_days row, fetch the tweets for the given day by querying
+    the list of tweets from CH and then fetching each tweet from twitter.
+
+    Arguments:
+    db - db handle
+    twitter_class - AbstractTwitter class
+    topic - topic dict
+    topic_tweet_day - topic_tweet_day dict
+    max_tweets - max tweets to fetch for a single day
+
+    Return:
+    None
+    """
+    if topic_tweet_day['tweets_fetched']:
+        return
+
+    ch_posts_data = topic_tweet_day['ch_posts']
+
+    ch_posts = ch_posts_data['posts']
+
+    if (max_tweets is not None):
+        ch_posts = ch_posts[0:max_tweets]
+
+    logger.debug("adding %s tweets for topic %s, day %s" % (len(ch_posts), topic['topics_id'], topic_tweet_day['day']))
+
+    # we can only get 100 posts at a time from twitter
+    for i in range(0, len(ch_posts), 100):
+        _add_tweets_to_ch_posts(twitter_class, ch_posts[i:i+100])
+
+    db.begin()
+
+    logger.debug("inserting into topic_tweets ...")
+
+    for ch_post in ch_posts:
+        if 'tweet' in ch_post:
+            _store_tweet_and_urls(db, topic, topic_tweet_day, ch_post)
+
+    num_deleted_tweets = len(list(filter(lambda x: 'tweet' not in x, ch_posts)))
+    topic_tweet_day['num_ch_tweets'] -= num_deleted_tweets
+
+    db.query(
+        "update topic_tweet_days set tweets_fetched = true, num_ch_tweets = $2 where topic_tweet_days_id = $1",
+        topic_tweet_day['topic_tweet_days_id'],
+        topic_tweet_day['num_ch_tweets'])
+
+    db.commit()
+
+    logger.debug("done inserting into topic_tweets")
+
+
+def _add_topic_tweet_single_day(
+        db: DatabaseHandler,
+        topic: dict,
+        day: datetime,
+        ch_class: AbstractCrimsonHexagon) -> dict:
+    """
+    Add a row to topic_tweet_day if it does not already exist.  fetch data for new row from CH.
+
+    Arguments:
+    db - database handle
+    topic - topic dict
+    day - date to fetch eg '2017-12-30'
+    ch_class - AbstractCrimsonHexagon class
+
+    Return:
+    None
+    """
+    topic_tweet_day = db.query(
+        "select * from topic_tweet_days where topics_id = $1 and day = $2",
```

Comment:
this is frustrating.  the ? and # parameters are not a dbix simple specific thing.  they are universal conventions for sql parameters.  it's ugly to repeat 'topics_id' and 'day' in three different places right next to each other.  I've changed to:

```
    db.query(
        "update topic_tweet_days set tweets_fetched = true, num_ch_tweets = %(a)s where topic_tweet_days_id = %(b)s",
        {'a': topic_tweet_day['num_ch_tweets'], 'b': topic_tweet_day['topic_tweet_days_id']})
```

Reply:
Thanks! Dunno if there a universal convention for parameter placeholders, but the reason to pass parameters like that is to be able to use psycopg's own parameter interpolation code instead of my awful "parameter translation" subroutines (which is rather stable but still brittle in many ways and probably has a security flaw or two somewhere).

Currently, one can pass parameters in following ways:

1. `db.query('SELECT * FROM a WHERE user_id = ? AND name = ?', user_id, name)` -- DBD::Pg's style 1; convenient, but uses brittle parameter translation code
2. `db.query('SELECT * FROM a WHERE user_id = $1 AND name = $2', user_id, name)` -- DBD::Pg's style 2; convenient, but uses brittle parameter translation code
3. `db.query('SELECT * FROM a WHERE user_id = %s AND name = %s', (user_id, name,))` -- psycopg's `tuple`; no awful brittle parameter translation is being done but `db` might be coming from Perl which has no idea what tuples are
4. `db.query('SELECT * FROM a WHERE user_id = %(user_id)s AND name = %(name)s', {'user_id': user_id, 'name': name})` -- psycopg's `dict`; very verbose, but no awful brittle parameter translation is being done and `dict` parameters get translated to Perl's hashrefs so everyone is happy

So, I'd suggest for us to do approach number 4 for the time being so that later I could make Media Cloud stop doing parameter interpolation itself and make it a psycopg's task.